File: sdks/rust/examples/rpc_quickstart.rs
//! RPC quickstart example for the Vectorizer Rust SDK (v3.x default).
//!
//! Connects to a server on `127.0.0.1:15503` (the default RPC port),
//! does the HELLO handshake, lists collections, and runs a search
//! against the first one.
//!
//! Run a Vectorizer server with RPC enabled (the v3.x default config
//! does this automatically), then:
//!
//! ```bash
//! cargo run --example rpc_quickstart
//! ```

use vectorizer_sdk::rpc::{HelloPayload, RpcClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Two equivalent ways to dial:
    //   1. By host:port (no scheme).
    //   2. By the canonical vectorizer:// URL.
    // The canonical URL is the recommended form because it makes the
    // transport explicit and round-trips through configuration files
    // unchanged.
    let url = std::env::var("VECTORIZER_URL")
        .unwrap_or_else(|_| "vectorizer://127.0.0.1:15503".to_string());
    println!("→ Dialing {url}");
    let client = RpcClient::connect_url(&url).await?;

    // HELLO handshake. In single-user mode (auth.enabled: false on
    // the server side), credentials are accepted-but-ignored. When
    // auth is enabled, attach a JWT or API key:
    //   HelloPayload::new("rpc-quickstart").with_token("<jwt>")
    let hello = client.hello(HelloPayload::new("rpc-quickstart")).await?;
    println!(
        "✓ HELLO ok — server={}  protocol_version={}  authenticated={}  admin={}",
        hello.server_version, hello.protocol_version, hello.authenticated, hello.admin
    );
    println!("  capabilities: {:?}", hello.capabilities);

    // PING (auth-exempt — works pre-HELLO too).
    let pong = client.ping().await?;
    println!("✓ PING → {pong}");

    // List collections.
    let collections = client.list_collections().await?;
    println!("✓ {} collection(s): {:?}", collections.len(), collections);

    // If we have at least one collection, run a search against it.
    if let Some(first) = collections.first() {
        println!("→ Searching '{first}' for 'vector database'");
        let info = client.get_collection_info(first).await?;
        println!(
            "  collection has {} vectors across {} documents (dim={})",
            info.vector_count, info.document_count, info.dimension
        );

        let hits = client.search_basic(first, "vector database", 5).await?;
        println!("  top {} hit(s):", hits.len());
        for hit in &hits {
            println!("    {} (score={:.4})", hit.id, hit.score);
        }
    } else {
        println!("  (no collections to search — create one via the REST/MCP API or the dashboard)");
    }

    Ok(())
}

File: sdks/python/examples/rpc_quickstart.py
"""RPC quickstart example for the Vectorizer Python SDK (v3.x default).

Connects to a server on ``127.0.0.1:15503`` (the default RPC port),
does the HELLO handshake, lists collections, and runs a search against
the first one.

Run a Vectorizer server with RPC enabled (the v3.x default config does
this automatically), then::

    cd sdks/python
    python -m examples.rpc_quickstart

Or, with a custom URL::

    VECTORIZER_URL=vectorizer://my-host:15503 python -m examples.rpc_quickstart
"""

from __future__ import annotations

import asyncio
import os
import sys

# Make the SDK importable when running this script directly from the
# checkout (without ``pip install -e .``). Real users who installed the
# package via pip can drop this prelude.
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from rpc import AsyncRpcClient, HelloPayload  # noqa: E402


async def main() -> None:
    # Two equivalent ways to dial:
    #   1. ``vectorizer://host:port`` — the canonical URL form. This
    #      is the recommended choice because it makes the transport
    #      explicit and round-trips through configuration files
    #      unchanged.
    #   2. Bare ``host:port`` (no scheme) — also accepted, treated as
    #      RPC.
    url = os.environ.get("VECTORIZER_URL", "vectorizer://127.0.0.1:15503")
    print(f"→ Dialing {url}")
    client = await AsyncRpcClient.connect_url(url)

    try:
        # HELLO handshake. In single-user mode (``auth.enabled: false``
        # on the server side), credentials are accepted-but-ignored.
        # When auth is enabled, attach a JWT or API key:
        #     HelloPayload(client_name="rpc-quickstart").with_token("<jwt>")
        hello = await client.hello(HelloPayload(client_name="rpc-quickstart"))
        print(
            f"✓ HELLO ok — server={hello.server_version}  "
            f"protocol_version={hello.protocol_version}  "
            f"authenticated={hello.authenticated}  admin={hello.admin}"
        )
        print(f"  capabilities: {hello.capabilities}")

        # PING (auth-exempt — works pre-HELLO too).
        pong = await client.ping()
        print(f"✓ PING → {pong}")

        # List collections.
        collections = await client.list_collections()
        print(f"✓ {len(collections)} collection(s): {collections}")

        # If we have at least one collection, run a search against it.
        if collections:
            first = collections[0]
            print(f"→ Searching '{first}' for 'vector database'")

            info = await client.get_collection_info(first)
            print(
                f"  collection has {info.vector_count} vectors across "
                f"{info.document_count} documents (dim={info.dimension})"
            )

            hits = await client.search_basic(first, "vector database", 5)
            print(f"  top {len(hits)} hit(s):")
            for hit in hits:
                print(f"    {hit.id} (score={hit.score:.4f})")
        else:
            print("  (no collections to search — create one via REST/MCP or the dashboard)")
    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(main())
